KAFKA-16505: Add source raw key and value

loicgreffier opened the pull request:
This PR is part of the
[KIP-1034](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1034%3A+Dead+letter+queue+in+Kafka+Streams).

It brings the support for [the source raw key and the source raw
value](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1034%3A+Dead+letter+queue+in+Kafka+Streams#KIP1034:DeadletterqueueinKafkaStreams-ErrorHandlerContext.java)
in the `ErrorHandlerContext`. Required by the routing to DLQ implemented
by https://github.com/apache/kafka/pull/17942.

Reviewers: Bruno Cadonna <cadonna@apache.org>

Dabz committed:
KAFKA-16505: Add source raw key and value

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -259,6 +259,10 @@ public <K, V> void send(final String topic,
 
+        // As many records could be in-flight,
+        // freeing raw records in the context to reduce memory pressure
+        freeContext(context);
```
The name of this method is a bit misleading. It basically frees the raw record within the context, not the whole context. What about calling it `freeRawInputRecordFromContext()`?

loicgreffier replied:
Fixed

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StampedRecord.java:
```diff
@@ -55,8 +69,26 @@ public Headers headers() {
     public String toString() {
         return value.toString() + ", timestamp = " + timestamp;
     }
+
+    @Override
+    public boolean equals(final Object other) {
+        return super.equals(other);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
+    }
```
Why are those needed?

loicgreffier replied:
Since we added new `rawKey` and `rawValue` attributes, SpotBugs requires to define the `equals` function (https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#eq-class-doesn-t-override-equals-in-superclass-eq-doesnt-override-equals)

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -48,6 +50,24 @@ public ProcessorRecordContext(final long timestamp,
         this.topic = topic;
         this.partition = partition;
         this.headers = Objects.requireNonNull(headers);
+        this.sourceRawKey = null;
+        this.sourceRawValue = null;
```
You also need to add these info to the `serialize()` and `deserialize()` so that the buffer values also get the source record. Here it gets a bit tricky, because you need to consider the case where a serialized record context does not contain the source record because it was written by a version of Streams that has not yet had the source record in the context.

loicgreffier replied:
Indeed having "optional" raw key and value make the deserialization tricky. 

Let's say we serialize the `ProcessorRecordContext` in this order `timestamp, offset, topic, partition, headers, rawKey, rawValue`. After deserializing the headers, the next bytes can be `rawKey` and `rawValue` or can be something else (e.g., `priorValue` https://github.com/apache/kafka/blob/d7a5b877f299ac20564dbb605e1a9c60fa835132/streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java#L73)

Right now I'm considering serializing the `rawKey` and `rawValue` at the very end of the `ByteBuffer` (i.e., right after here: https://github.com/apache/kafka/blob/d7a5b877f299ac20564dbb605e1a9c60fa835132/streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java#L119). Thus, after deserializing all the non-optional fields if there is some bytes remaining in the buffer, it should be the `rawKey` and `rawValue`.

cadonna reviewed (commented):
Thanks @loicgreffier for the PR!

You did not consider buffers in this PR as I described in [my comment](https://github.com/apache/kafka/pull/16684#issuecomment-2249925420). Could you come up with a test that confirms that we also have the same issue with buffers and then provide a fix?

loicgreffier commented:
@cadonna Changes about buffers will be added to this PR. However, despite my tests using `suppress()`, I did not manage to lose the `rawKey` and `rawValue` for now. I'm always receiving a value for these fields in the `processingExceptionHandler`.

cadonna commented:
> @cadonna Changes about buffers will be added to this PR. However, despite my tests using `suppress()`, I did not manage to lose the `rawKey` and `rawValue` for now. I'm always receiving a value for these fields in the `processingExceptionHandler`.

I believe, your test needs to flush the buffer so that the records are written to the changelog topic and then restore the buffer from the changelog topic by stopping and re-starting the app.

Useful code:
- The record is written to the set of dirty keys: https://github.com/apache/kafka/blob/291523e3e40567f961b4d94068e6946d751d18e4/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java#L489
- The set of dirty keys is flushed to the changelog: https://github.com/apache/kafka/blob/291523e3e40567f961b4d94068e6946d751d18e4/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java#L247
- The buffer is restored from the changelog: https://github.com/apache/kafka/blob/291523e3e40567f961b4d94068e6946d751d18e4/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueChangeBuffer.java#L300

loicgreffier commented:
@cadonna Thank you for the guidance. I could trigger the `InMemoryTimeOrderedKeyValueChangeBuffer#flush` and `InMemoryTimeOrderedKeyValueChangeBuffer#restoreBatch` and confirm that the `rawKey` and `rawValue` are lost.

Please ignore my previous comment https://github.com/apache/kafka/pull/18739#discussion_r1937172436 that brings too many changes.

I've updated the serialization and deserialization. To take into consideration the optional rawKey and rawValue, I've added a char "marker" (i.e., `k` for rawKey, `v` for rawValue) to identify the presence of these values without mixing with any other possible bytes.

Let me know your thoughts about this approach

cadonna commented:
@loicgreffier I discussed the need of writing raw key and raw value of the input record to the changelog topic for buffers with @mjsax and we had some concerns. Writing the input record to the changelog topic might significantly increase the storage requirements because we would need to write two records for each record in the buffer, the record itself and the corresponding input record.

loicgreffier commented:
@cadonna I understand the concern. Should we restart the thread and discuss the possible alternatives? 
- Make the sourceRawKey/sourceRawValue a conscious trade against memory with a new configuration?
- Drop the sourceRawKey/sourceRawValue and use the current record instead?

cadonna commented:
@loicgreffier Sorry for the late reply!
Coincidentally, I talked to a user of Kafka Streams that implemented a dead letter queue. They have the requirement to write the raw input record to the dead letter queue. That means, we have two data point from users (yours and this user) that agree on the need of the input record in the dead letter queue.

The issue with the raw input record described above only applies to Streams applications that use suppress or stream-table joins with versioned tables and grace periods. IMO, we should not exclude the raw input record from the error handlers because of some applications. One option can be that writing the raw input record to the changelog topic is enabled by a configuration as you propose above. That means that error handlers need to gracefully handle the situation where the raw input record is not available.

WDYT?

loicgreffier commented:
@cadonna Agree to introduce a new parameter. 

I'm wondering if the concern is the same for store caches. Should the parameter applies to store caches as well (`CachingKeyValueStore`, `CachingSessionStore`, `CachingWindowStore`) ? Or should we consider that there is no impact at all?

cadonna commented:
@loicgreffier 

The first question we need to answer is the following:
Should the raw input record be (a) the record from which the defective record that triggered the handler was derived from or should the raw input record be (b) the record that triggered the defective record to be sent downstreams?

If we decide for (a), we need to define the input record for the DSL operation. For example, what is the input record that should be written to the dead letter queue in case an aggregation result triggers the error handler? Intuitively, I would say it is the last record that contributed to the aggregation. However, there are cases where this might not be true. For example, if the aggregate is a list of accumulated records and the last record is not responsible for the error downstreams. With (a), ideally, we need to carry the raw input record everywhere, into state stores, changelogs, caches, and buffers.  
I checked the code. I believe, we have an issue also with caches since I realized that when an entry is added to the cache during a get operation, the record context is populated with sentinel values like `-1` and `null`:
https://github.com/apache/kafka/blob/f13a22af0b3a48a4ca1bf2ece5b58f31e3b26b7d/streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java#L372.      

If we decide for (b), maintaining the raw input record might be a bit simpler but I am not sure how useful the raw input record is in this case. You cannot really understand why something went wrong and replay.

Let me know what you think.

loicgreffier commented:
@cadonna KIP-1034 suggests that the input record should be the one that triggers the sub-topology: https://cwiki.apache.org/confluence/display/KAFKA/KIP-1034%3A+Dead+letter+queue+in+Kafka+Streams#KIP1034:DeadletterqueueinKafkaStreams-DefaultDeadletterqueuerecord

> (a) the record from which the defective record that triggered the handler was derived from

(a) aligns with the definition in KIP-1034, doesn't it? It makes it easier to reprocess records.

> (b) the record that triggered the defective record to be sent downstreams?

Does this mean that in the following example:

```java
.stream()
.selectKey() 
.mapValues() <----------- Exception here
...
```

The record sent to the DLQ would be the input record of the mapValues processor? (i.e., the one passed to the processing error handler?

---

> Intuitively, I would say it is the last record that contributed to the aggregation.

```java
.stream(INPUT_TOPIC)
.selectKey()
.groupByKey()
.aggregate(
  initializer,
  aggregator, <----------- Exception here
  materialized
)
...
```

With the current implementation:
- The source record is the last record that contributed to the aggregation.
- The source topic is the repartition topic.
- The default DLQ topic, specified by `errors.dead.letter.queue.topic.name`, may contain records from different sub-topologies (e.g., records from `INPUT_TOPIC` or the `repartition` topic).

---

> I believe, we have an issue also with caches since I realized that when an entry is added to the cache during a get operation

I will check and reproduce this case

@sebastienviale Feel free to add anything if I missed a point

sebastienviale commented:
As @loicgreffier said, it is mentioned in the KIP that the raw record value: "If available, contains the value of the input message that triggered the sub-topology, null if triggered by punctuate"

Which seems to correspond to solution (a). It also seems logical to store in the raw record the last record before the exception occurs.

loicgreffier commented:
Some tests done:

## ✅ Stateless

```java
streamsBuilder
  .stream()
  .peek()
  .selectKey()
  .mapValues() // <----- Exception 💥
  ```

- Source raw key and source raw value from the input topic.

## ✅ Aggregate Caching Enabled (Exception in Aggregator)

```java
streamsBuilder
  .stream()
  .peek()
  .groupBy()
  .aggregate(
    () -> new KafkaUserAggregate(new ArrayList<>()),
    new UserAggregator(), // <----- Exception 💥
    Materialized.<String, KafkaUserAggregate, KeyValueStore<Bytes, byte[]>>as(USER_AGGREGATE_STORE)
        .withKeySerde(Serdes.String())
        .withValueSerde(SerdesUtils.getValueSerdes())
        .withCachingEnabled());
```

- Source raw key and source raw value from the repartition topic.
- Source raw key and source raw value of the latest record that enters the aggregator.

## ✅ Aggregate Caching Enabled (Exception after aggregation)

```java
streamsBuilder
  .stream()
  .peek()
  .groupByKey()
  .aggregate(
    () -> new KafkaUserAggregate(new ArrayList<>()),
    new UserAggregator(),
    Materialized.<String, KafkaUserAggregate, KeyValueStore<Bytes, byte[]>>as(USER_AGGREGATE_STORE)
        .withKeySerde(Serdes.String())
        .withValueSerde(SerdesUtils.getValueSerdes())
        .withCachingEnabled())
  .mapValues() // <----- Exception 💥
```

- Source raw key and source raw value from the repartition topic.
- Source raw key and source raw value of the latest record that enters the aggregator.
- ⚠️ Needs the updates on `CachingKeyValueStore.java`, `CachingSessionStore.java`, `CachingWindowStore.java`. Otherwise,
source raw key and source raw value are null.

## ✅ Table Caching Enabled

```java
streamsBuilder
  .table(
    USER_TOPIC,
    Consumed.with(Serdes.String(), SerdesUtils.getValueSerdes()),
    Materialized.<String, KafkaUser, KeyValueStore<Bytes, byte[]>>as("USER_STORE")
        .withKeySerde(Serdes.String())
        .withValueSerde(SerdesUtils.getValueSerdes())
        .withCachingEnabled())
  .mapValues(); // <----- Exception 💥
```

- Source raw key and source raw value from the input topic.
- ⚠️ Needs the updates on `CachingKeyValueStore.java`, `CachingSessionStore.java`, `CachingWindowStore.java`. Otherwise,
source raw key and source raw value are null (like above).

## ✅ Left Join Stream-Stream

```java
streamsBuilder
  .stream()
  .selectKey()
  .leftJoin(streamTwo, (key, left, right) -> { ... // <----- Exception 💥});
```
- Source raw key and source raw value from the left repartition topic.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:
```diff
@@ -147,4 +147,38 @@ public interface ErrorHandlerContext {
+     *
+     * <p> If this method is invoked within a {@link Punctuator#punctuate(long)
+     * punctuation callback}, or while processing a record that was forwarded by a punctuation
+     * callback, it will return null.
```
nit:
```suggestion
     * callback, it will return {@code null}.
```

loicgreffier replied:
Fixed

cadonna commented on streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:
```diff
@@ -147,4 +147,38 @@ public interface ErrorHandlerContext {
+     *
+     * <p> If this method is invoked within a {@link Punctuator#punctuate(long)
+     * punctuation callback}, or while processing a record that was forwarded by a punctuation
+     * callback, it will return null.
```
nit:
```suggestion
     * callback, it will return {@code null}.
```

loicgreffier replied:
Fixed

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/RecordContext.java:
```diff
@@ -110,4 +110,37 @@ public interface RecordContext {
+     *
+     * <p> If this method is invoked within a {@link Punctuator#punctuate(long)
+     * punctuation callback}, or while processing a record that was forwarded by a punctuation
+     * callback, it will return null.
```
nit:
```suggestion
     * callback, it will return {@code null}.
```

loicgreffier replied:
Fixed

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/RecordContext.java:
```diff
@@ -110,4 +110,37 @@ public interface RecordContext {
+     *
+     * <p> If this method is invoked within a {@link Punctuator#punctuate(long)
+     * punctuation callback}, or while processing a record that was forwarded by a punctuation
+     * callback, it will return null.
```
nit:
```suggestion
     * callback, it will return {@code null}.
```

loicgreffier replied:
Fixed

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/RecordContext.java:
```diff
@@ -110,4 +110,37 @@ public interface RecordContext {
+     * to the repartition topic.
+     *
+     * <p> Always returns null if this method is invoked within a
+     * ProductionExceptionHandler.handle(ErrorHandlerContext, ProducerRecord, Exception)
```
Is this method ever called in `ProductionExceptionHandler.handle()`? Is this a copy&paste mistake from `ErrorHandlerContext`?

loicgreffier replied:
Only `ErrorHandlerContext` is given to the `ProductionExceptionHandler#handle`, so this is a bad c&p. I've updated both code and KIP.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/RecordContext.java:
```diff
@@ -110,4 +110,37 @@ public interface RecordContext {
+     * to the repartition topic.
+     *
+     * <p> Always returns null if this method is invoked within a
+     * ProductionExceptionHandler.handle(ErrorHandlerContext, ProducerRecord, Exception)
```
Is this method ever called in `ProductionExceptionHandler.handle()`? Is this a copy&paste mistake from `ErrorHandlerContext`?

loicgreffier replied:
See https://github.com/apache/kafka/pull/18739#discussion_r2125010510

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:
```diff
@@ -1890,6 +1892,69 @@ public void shouldNotSendIfSendOfOtherTaskFailedInCallback() {
+        );
+
+        collector.send(topic, "3", "0", null, null, stringSerializer, stringSerializer, sinkNodeName, context, streamPartitioner);
+        collector.flush();
```
Why do you flush before the assertion?
Could you assert before flush?

loicgreffier replied:
Flush was not necessary in the test, it has been removed.

cadonna reviewed (commented):
Thanks for the updates @loicgreffier !

Here my feedback.

We are almost there. 

Could you add tests (or maybe only additional assertions in existing tests are needed) for the cases you pointed out above?

loicgreffier commented:
@cadonna 

- I've mainly updated `ProcessingExceptionHandlerIntegrationTest.java` to assert the presence of the source raw key and source raw value. 
- I’ve added a new parameterized test `shouldVerifySourceRawKeyAndSourceRawValuePresentOrNotInErrorHandlerContext` that verifies whether these fields are correctly set in the error handler context across a variety of topologies.

Please let me know if you think this approach is not the appropriate way to test it, or if you believe additional test cases should be included.

cadonna reviewed (approved):
@loicgreffier Thanks for the updates!

LGTM!

cadonna commented:
@loicgreffier Could you please a description to the PR and remove the template text?

loicgreffier commented:
@cadonna Done. I've updated the description.
